pkgs/os-specific/linux/kernel/cachyos-kernels.nix

```
@@ -0,0 +1,155 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  pkgs,
+  buildLinux,
+  ...
+}@args:
+
+let
+  cachyPatches = fetchFromGitHub {
+    owner = "CachyOS";
+    repo = "kernel-patches";
+    rev = "04fb9178158a7714aeb908c7a5310c4e0d6ea4b6";
+    hash = "sha256-KTvZxYy+UgLNg1DYBMrJrRdxDo3N5ususlrHyRBS7C8=";
+  };
+
+  ltsVariant = {
+    version = "6.6.43";
+    source = {
+      url = "mirror://kernel/linux/kernel/v6.x/linux-6.6.43.tar.xz";
+      hash = "sha256-Ctg7Ghp4ChqtlI1VqlXuY8UMYm8tRpELnSGAAo0QCl4=";
+    };
+    patches = [
+      {
+        name = "cachyos-base-all";
+        patch = "${cachyPatches}/6.6/all/0001-cachyos-base-all.patch";
+      }
+      {
+        name = "bore-cachy";
+        patch = "${cachyPatches}/6.6/sched/0001-bore-cachy.patch";
+      }
+    ];
+
+    config = with lib.kernel; {
+
+    };
+  };
+
+  mainVariant = {
+    version = "6.10.2";
+    source = {
+      url = "mirror://kernel/linux/kernel/v6.x/linux-6.10.2.tar.xz";
+      hash = "sha256-c9hSDdnLpaz8XnII52s12XQLiq44IQqSJOMuxMDSm3A=";
+    };
+    patches = [
+      {
+        name = "cachyos-base-all";
+        patch = "${cachyPatches}/6.10/all/0001-cachyos-base-all.patch";
+      }
+      {
+        name = "sched-ext";
+        patch = "${cachyPatches}/6.10/sched/0001-sched-ext.patch";
+      }
+      {
+        name = "bore-cachy-ext";
+        patch = "${cachyPatches}/6.10/sched/0001-bore-cachy-ext.patch";
+      }
+    ];
+
+    config = with lib.kernel; {
+      # CachyOS config
+      CACHY = yes;
+
+      # CPU Scheduler
+      SCHED_CLASS_EXT = yes;
+      SCHED_BORE = yes;
+      MIN_BASE_SLICE_NS = 1000000;
+
+      # LLVM level
+      LTO_NONE = yes;
+
+      # Tick rate
+      HZ_300 = null;
+      HZ_1000 = yes;
+      HZ = 1000;
+
+      NR_CPUS = 320;
+
+      # Tick type
+      HZ_PERIODIC = null;
+      NO_HZ_IDLE = null;
+      CONTEXT_TRACKING_FORCE = null;
+      NO_HZ_FULL_NODEF = yes;
+      NO_HZ_FULL = yes;
+      NO_HZ = yes;
+      NO_HZ_COMMON = yes;
+      CONTEXT_TRACKING = yes;
+
```

The cachy kernel has some RCU subsystem tweaks. They're mostly the same as in the Zen kernel which is already in nixpkgs, so you can just copy those. The only difference is that cachy unsets `RCU_NOCB_CBU_{DEFAULT_ALL, CB_BOOST}` and enables `RCU_LAZY_DEFAULT_OFF`.

```suggestion

      # RCU
      TREE_RCU = yes;
      PREEMPT_RCU = yes;
      RCU_EXPERT = yes;
      TREE_SRCU = yes;
      TASKS_RCU_GENERIC = yes;
      TASKS_RCU = yes;
      TASKS_RUDE_RCU = yes;
      TASKS_TRACE_RCU = yes;
      RCU_STALL_COMMON = yes;
      RCU_NEED_SEGCBLIST = yes;
      RCU_FANOUT = freeform "64";
      RCU_FANOUT_LEAF = freeform "16";
      RCU_BOOST = yes;
      RCU_BOOST_DELAY = option (freeform "500");
      RCU_NOCB_CPU = yes;
      RCU_LAZY = yes;
      RCU_DOUBLE_CHECK_CB_TIME = yes;
      RCU_NOCB_CPU_DEFAULT_ALL = unset;
      RCU_NOCB_CPU_CB_BOOST = unset;
      RCU_LAZY_DEFAULT_OFF = yes;
```

---

I don't see these in the script, are these an Arch default?

---

You're right, these are currently an Arch default: https://gitlab.archlinux.org/archlinux/packaging/packages/linux/-/blob/main/config#L163-189

I thought these settings could be a CachyOS-specific tweak because common for these desktop performance oriented kernels to mess with RCU settings, for example both Zen and Xanmod set stuff similar to this and they aren't exactly Arch-specific. The commits in CachyOS/linux-cachyos which introduced these changes are from 2021-2022 and are poorly described, so it's hard to tell if the change came from Arch or was it a cachy customization back then.  

These RCU settings can affect latency performance, so they might be worth including.

---

To be honest I'm hesitant to deviate too far from the NixOS defaults when Cachy doesn't explicitly do things differently, but I guess I can reach out to someone from their project and see what they think.

---

https://github.com/CachyOS/linux-cachyos/discussions/273